pubsub/psiface/interfaces.go

```diff
@@ -0,0 +1,51 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package psiface
+
+import (
+	"context"
+
+	"cloud.google.com/go/pubsub"
+)
+
+type Client interface {
+	CreateTopic(ctx context.Context, topicID string) (Topic, error)
+	Topic(id string) Topic
+	CreateSubscription(ctx context.Context, id string, cfg SubscriptionConfig) (Subscription, error)
+	Subscription(id string) Subscription
+
+	embedToIncludeNewMethods()
+}
+
+type Topic interface {
+	String() string
+	Publish(ctx context.Context, msg *pubsub.Message) PublishResult
+
+	embedToIncludeNewMethods()
+}
+
+type Subscription interface {
+	Exists(ctx context.Context) (bool, error)
+	Receive(ctx context.Context, f func(context.Context, *pubsub.Message)) error
```

Comment:
I think this should take a Message we define here, instead of a *pubsub.Message.

(ditto elsewhere)

This is because Ack/Nack will fail when we use an artificial pubsub.Message, and also because Ack/Nack are things people will most likely want to mock.

Reply:
Sounds good to me. I've actually written code that uses reflection to provide implementations of `Ack()` and `Nack()`, which is pretty ugly. If we use `Message` for the interface, any preferences for how to obtain an implementation?

I can see two choices: either make the struct definition public (`DefaultMessage` perhaps?), or add an `AdaptMessage` method. I'd probably lean towards the latter, but thought I'd check as there doesn't seem to be anywhere in the existing code where multiple adapters are public.

Reply:
I've pushed a few more commits implementing that approach. Note that if I was creating a fake `Message` for my own purposes I would probably use functional options on the constructor rather than a long list of parameters, but doing so here probably obfuscates the example and so I've stuck with the burdensome signature:

```golang
func newFakeMessage(id string, data []byte, attributes map[string]string, publishTime time.Time) *fakeMessage
```